File: spring_java/src/main/java/dev/pollito/spring_java/config/security/handler/CustomAccessDeniedHandler.java
package dev.pollito.spring_java.config.security.handler;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import org.jspecify.annotations.NonNull;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

@Component
public class CustomAccessDeniedHandler implements AccessDeniedHandler {

  @Override
  public void handle(
      @NonNull HttpServletRequest request,
      @NonNull HttpServletResponse response,
      @NonNull AccessDeniedException accessDeniedException)
      throws IOException {
    AuthenticationErrorResponseWriter.write(
        request, response, HttpStatus.FORBIDDEN, accessDeniedException.getMessage());
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/config/security/handler/CustomAccessDeniedHandler.kt
package dev.pollito.spring_kotlin.config.security.handler

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus.FORBIDDEN
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component

@Component
class CustomAccessDeniedHandler : AccessDeniedHandler {

  override fun handle(
      request: HttpServletRequest,
      response: HttpServletResponse,
      accessDeniedException: AccessDeniedException,
  ) {
    AuthenticationErrorResponseWriter.write(
        request,
        response,
        FORBIDDEN,
        accessDeniedException.message ?: "Forbidden",
    )
  }
}
